semgrep/semgrep/core_runner.py

```diff
@@ -150,11 +150,16 @@ def _flatten_rule_patterns(self, rules: List[Rule]) -> Iterator[Pattern]:
     def _group_patterns_by_language(
         self, rules: List[Rule]
     ) -> Dict[Language, List[Pattern]]:
-        # a rule can have multiple patterns inside it. Flatten these so we can send semgrep a single yml file list of patterns
-        patterns: List[Pattern] = list(self._flatten_rule_patterns(rules))
+
         by_lang: Dict[Language, List[Pattern]] = collections.defaultdict(list)
-        for pattern in patterns:
-            by_lang[pattern.language].append(pattern)
+        if rules[0]._mode == "taint":
```

Comment:
`[0]` always makes me nervous because what if `rules` is empty or there are mixed rules and you only consider the first case. I would generally tend to use map, filter, partition logic here.

Reply:
re-factored it so this isn't even a list (it was always [rule] before, and doesn't need to become a list until later)